src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+
+/**
+ * Represents a string.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    protected $string = '';
+    protected $locale = null;
+    protected $ignoreCase = false;
+
+    abstract public function __construct(string $string = '', string $locale = null);
+
+    /**
+     * @return static
+     */
+    abstract public function after(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function afterLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function append(string ...$suffix): self;
+
+    /**
+     * @return static
+     */
+    abstract public function before(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static
+     */
+    abstract public function beforeLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
+    {
+        $str = clone $this;
+        $str->string = trim(preg_replace('/(?:\s{2,}+|[^\S ])/', ' ', $str->string));
+
+        return $str;
+    }
+
+    abstract public function endsWith(string $suffix): bool;
+
+    /**
+     * @return static
+     */
+    public function ensureEnd(string $suffix): self
+    {
+        return $this->endsWith($suffix) ? clone $this : $this->append($suffix);
+    }
+
+    /**
+     * @return static
+     */
+    public function ensureStart(string $prefix): self
+    {
+        return $this->startsWith($prefix) ? clone $this : $this->prepend($prefix);
+    }
+
+    abstract public function equalsTo(string $string): bool;
+
+    /**
+     * @return static
+     */
+    abstract public function folded(): self;
+
+    /**
+     * @return static
+     */
+    public function ignoreCase(): self
```

Comment:
> duplicating all methods with "IgnoreCase" variants

@nicolas-grekas For the record, only 1/4 of methods would need it, but I agree it's already a lot.

> making case sensitivity as a state [...] would be a wrong model for strings

It's already the case, isn't it ? Only temporary until the next method call. It's this behaviour that surprised me (*a method that alters the way the very next method works*), I don't recall having seen it elsewhere. Why not `->includeNeedle()` instead of a second parameter `*(string $needle, bool $includeNeedle = false)` in the 4 methods using this parameter, then ?

Judjing from @stof comment "*no way to unignore case ?*", I'm probably not the only one surprised. I'm adding 2 suggestions, and I won't insist after that ;) 

5) `->IgnoreCaseNext()` : to make it explicit that only the next call is affected

6) `->caseSensitive()` and `->caseInsensitive()`, without the "reset state" feature in `__clone`, to make the state more explicit, as it already exists anyway. Same suggestion as *4)* but without the constructor parameter.

My favourite solution was *2)* : `, $caseSensitive = true)` to stay consistent with `, bool $includeNeedle = false)`

Thanks for your work anyway :tada: